DEPR: Deprecate tshift and integrate it to shift

fujiaxiang opened the pull request:
- [x] closes #11631
- xref #34452
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

fujiaxiang committed:
deprecated tshift and integrated it to shift (GH11631)

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into deprecate_tshift

fujiaxiang committed:
added deprecated sphinx directive

fujiaxiang committed:
removed mention of tshift in timeseries user guide

fujiaxiang committed:
adjusted tests, added filterwarnings mark where applicable

jreback commented:
thanks. would not be averse to actually re-writing tshift in terms of shift; I *think* this would be straightforward. Also could merge this and then followon to do that.

jreback commented:
> > would not be averse to actually re-writing tshift in terms of shift; I think this would be straightforward
> 
> Agree it's rather straightforward, but just to clarify a few things:
> 
> 1. we are still going to deprecate `tshift` after the re-write right? In other words, you don't mean to re-write and **keep** `tshift` right?

correct, still deprecate

> 2. do you think it's better to move the current `_tshift` inside `shift`? i.e. we won't have a separate method `_tshift`, but unpack what's in it and put inside `shift`

yes if possible (ideally simplified as much as possible).

fujiaxiang committed:
refactored and simplified shift and tshift

fujiaxiang commented:
Have unpacked previous `_tshift` and moved into `shift`, also simplified the code a little bit. The deprecating `tshift` now calls `shift`. @jreback can you review?

jreback commented on pandas/tests/frame/methods/test_shift.py:
```diff
@@ -186,10 +189,60 @@ def test_tshift(self, datetime_frame):
+        shifted2 = ps.shift(freq="B")
+        tm.assert_frame_equal(shifted, shifted2)
+
+        shifted3 = ps.shift(freq=offsets.BDay())
```
can you put the failures in a separate test (e.g. things that assert raises, name it like test_shift_with_freq_errors

also not averse to separating PI and DTI into 2 tests. if possibly to parameterize would be great but if not ok too.

fujiaxiang replied:
done

jreback commented on pandas/tests/series/methods/test_shift.py:
```diff
@@ -220,10 +223,58 @@ def test_tshift(self, datetime_series):
+        with tm.assert_produces_warning(FutureWarning):
+            datetime_series.tshift()
+
+    def test_shift_with_freq(self, datetime_series):
```
same comment as above, e.g. to split this up a bit

fujiaxiang replied:
done

jreback commented:
@fujiaxiang can you rebase and some small comments, ping on green.

fujiaxiang committed:
Merge remote-tracking branch 'upstream/master' into deprecate_tshift

fujiaxiang committed:
logically split up some tests

fujiaxiang commented:
@jreback have split up the tests into PI and DTI and also separated the failure tests
I tried to parametrize the test cases into PI and DTI but I realize they only have a very small proportion of code in common, so I decided to separate them into different tests

jreback commented:
> @jreback have split up the tests into PI and DTI and also separated the failure tests
> I tried to parametrize the test cases into PI and DTI but I realize they only have a very small proportion of code in common, so I decided to separate them into different tests

kk sounds good. ping on green (note that there maybe be a failure on master that is unrelated)

jreback commented:
thanks @fujiaxiang very nice
